caledonia/src/bounded.rs

```diff
@@ -0,0 +1,451 @@
+    /// Returns an empty proof
+    fn empty() -> Self {
+        Proof {
+            r: 0,
+            d: 0,
+            items: Vec::new(),
+        }
+    }
+
+    /// Oracle producing a uniformly random value in [1, n_p] used for prehashing S_p
+    // TODO: We also return hash(data) to follow the optimization presented in Section 3.3
+    fn h0(setup: &Setup, v: usize, s: [u8; 32]) -> usize {
+        let mut data = Vec::new();
+        data.push(v.to_ne_bytes().to_vec());
+        data.push(s.to_vec());
+        let digest = utils::combine_hashes(data);
+        // return (digest, utils::oracle(&digest, setup.n_p));
+        return utils::oracle(&digest, setup.n_p);
+    }
+
+    /// Oracle defined as Bernoulli(q) returning 1 with probability q and 0 otherwise
+    fn h2(setup: &Setup, r: &Round) -> bool {
+        let mut data = Vec::new();
+        data.push(r.v.to_ne_bytes().to_vec());
+        data.push(r.t.to_ne_bytes().to_vec());
+        for s in &r.s_list {
+            data.push(s.clone().to_vec());
+        }
+        let digest = utils::combine_hashes(data);
+        return utils::oracle(&digest, setup.q) == 0;
+    }
+
+    /// Depth-first search which goes through all potential round candidates
+    /// and returns first round candidate Round{t, x_1, ..., x_u)} such that:
+    /// - for all i ∈ [0, u-1], H0(x_i+1) ∈ bins[H1(t, x_1, ..., x_i)]
+    /// - H2(t, x_0, ..., x_u) = true
+    fn dfs(
+        setup: &Setup,
+        bins: &Vec<Vec<[u8; 32]>>,
+        round: &Round,
+        limit: &mut usize,
+    ) -> Option<Proof> {
+        // If the round candidate is at the last iteration (len(round) == u)
+        if round.s_list.len() == setup.u {
+            // if H2(t, x_0, ..., x_u) = true, return candidate as proof
+            if Proof::h2(setup, round) {
+                let r = round.v;
+                let d = round.t;
+                let items = round.s_list.clone();
+                return Some(Proof { r, d, items });
+            } else {
+                return None;
+            }
+        }
+        // Otherwise, update round candidate with all s in bins[H1(t, x_1, ..., x_i)] and continue
+        let result = bins[round.h_usize].iter().find_map(|&s| {
+            if *limit == 0 {
+                return None;
+            }
+            *limit -= 1;
+            Self::dfs(setup, bins, &Round::update(round, s), limit)
+        });
+        return result;
+    }
+
+    /// Alba's proving algorithm, based on a depth-first search algorithm.
+    /// Returns an empty proof if no suitable candidate is found.
+    fn prove_index(setup: &Setup, set: &Vec<[u8; 32]>, v: usize) -> (usize, Option<Proof>) {
+        // Initialising our n_p bins with all s ∈ Sp
+        let mut bins: Vec<Vec<[u8; 32]>> = Vec::new();
+        for _ in 1..(setup.n_p + 1) {
+            bins.push(Vec::new());
+        }
+        for &s in set.iter() {
+            // TODO: add H1(s),
+            // let (hs, index_s) = Proof::h0(setup, s);
```

Comment:
Remove commented code (plus the `TODO` is misleading as it mentions `H1` and we use `h0`